src/main/java/com/google/devtools/build/lib/sandbox/WindowsSandboxedSpawnRunner.java

```diff
@@ -0,0 +1,145 @@
+import com.google.devtools.build.lib.util.OS;
+import com.google.devtools.build.lib.vfs.FileSystem;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import com.google.devtools.build.lib.vfs.Symlinks;
+import java.io.File;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
+import javax.annotation.Nullable;
+
+/** Spawn runner that uses BuildXL Sandbox APIs to execute a local subprocess. */
+final class WindowsSandboxedSpawnRunner extends AbstractSandboxSpawnRunner {
+
+  private final FileSystem fileSystem;
+  private final BlazeDirectories blazeDirs;
+  private final Path execRoot;
+  private final PathFragment windowsSandbox;
+  private final Path sandboxBase;
+  private final LocalEnvProvider localEnvProvider;
+  private final Duration timeoutKillDelay;
+  private final TreeDeleter treeDeleter;
+
+  /**
+   * Creates a sandboxed spawn runner that uses the {@code windows-sandbox} tool.
+   *
+   * @param cmdEnv the command environment to use
+   * @param sandboxBase path to the sandbox base directory
+   * @param timeoutKillDelay an additional grace period before killing timing out commands
+   */
+  WindowsSandboxedSpawnRunner(
+      CommandEnvironment cmdEnv,
+      Path sandboxBase,
+      Duration timeoutKillDelay,
+      PathFragment windowsSandboxPath,
+      TreeDeleter treeDeleter) {
+    super(cmdEnv);
+    this.fileSystem = cmdEnv.getRuntime().getFileSystem();
+    this.blazeDirs = cmdEnv.getDirectories();
+    this.execRoot = cmdEnv.getExecRoot();
+    this.windowsSandbox = windowsSandboxPath;
+    this.sandboxBase = sandboxBase;
+    this.timeoutKillDelay = timeoutKillDelay;
+    this.localEnvProvider = new WindowsLocalEnvProvider(cmdEnv.getClientEnv());
+    this.treeDeleter = treeDeleter;
+  }
+
+  @Override
+  protected SpawnResult actuallyExec(Spawn spawn, SpawnExecutionContext context)
+      throws IOException, ExecException, InterruptedException {
+    // Each invocation of "exec" gets its own sandbox base.
+    // Note that the value returned by context.getId() is only unique inside one given SpawnRunner,
+    // so we have to prefix our name to turn it into a globally unique value.
+    Path sandboxPath =
+        sandboxBase.getRelative(getName()).getRelative(Integer.toString(context.getId()));
+    sandboxPath.getParentDirectory().createDirectory();
+    sandboxPath.createDirectory();
+
+    // b/64689608: The execroot of the sandboxed process must end with the workspace name, just like
+    // the normal execroot does.
+    Path sandboxExecRoot = sandboxPath.getRelative("execroot").getRelative(execRoot.getBaseName());
+    sandboxExecRoot.getParentDirectory().createDirectory();
+    sandboxExecRoot.createDirectory();
+
+    Map<String, String> environment =
+        localEnvProvider.rewriteLocalEnv(spawn.getEnvironment(), binTools, "/tmp");
```

Comment:
`/tmp` should be a Windows path I think

Reply:
About this, I don't really know what is the good path I should set. `C:\Windows\Temp` is writable without admin permission, but not readable without admin permission. `%USERPROFILE%\AppData\Temp` probably exists, but `USERPROFILE` env var might not be present to construct the path.

In `LocalSpawnRunner`, it creates a unique tmp dir for this: https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/exec/local/LocalSpawnRunner.java;drc=c26e3396cb341e173b4be5c780cafac3ec386df5;l=310 . Should I do the same here?

Reply:
Don't use any of those. If neither TEMP nor TMP are set, things like `java.nio.file.Files.createTempDirectory` break anyway.

> Should I do the same here?

Yes please.